File: scripts/normalize.js
'use strict';

/**
 * normalize.js — Parse today's raw XML feed for each enabled source and write
 * normalized Opportunity records to data/normalized/opportunities/<today>.jsonl.
 *
 * Usage:
 *   node scripts/normalize.js
 */

const fs        = require('fs');
const path      = require('path');
const RssParser = require('rss-parser');

const SCRIPT_DIR = __dirname;
const REPO_ROOT  = path.resolve(SCRIPT_DIR, '..');

const { sha1Id, nowIso, todayStr, safeMkdir, writeJsonl } = require('./utils');

const SOURCES_FILE = path.join(REPO_ROOT, 'data', 'sources', 'sources.json');
const RAW_DIR      = path.join(REPO_ROOT, 'data', 'raw');
const NORM_DIR     = path.join(REPO_ROOT, 'data', 'normalized', 'opportunities');

// ---------------------------------------------------------------------------

function loadSources() {
  return JSON.parse(fs.readFileSync(SOURCES_FILE, 'utf8'));
}

/**
 * Strip HTML tags and decode common HTML entities from *text*.
 * @param {string|null|undefined} text
 * @returns {string|null}
 */
function cleanHtml(text) {
  if (!text) return null;
  const stripped = text
    .replace(/<[^>]+>/g, ' ')          // remove tags
    .replace(/&[a-zA-Z#\d]+;/g, ' ')  // remove entities (approx.)
    .replace(/\s+/g, ' ')
    .trim();
  return stripped || null;
}

/**
 * Return path to today's raw XML, falling back to latest.xml.
 * @param {string} source
 * @param {string} today   YYYY-MM-DD
 * @returns {string|null}
 */
function findRawFile(source, today) {
  const sourceDir = path.join(RAW_DIR, source);
  const daily     = path.join(sourceDir, `${today}.xml`);
  if (fs.existsSync(daily)) return daily;

  const latest = path.join(sourceDir, 'latest.xml');
  if (fs.existsSync(latest)) {
    console.log(`  ⚠ No daily file for ${today}, falling back to latest.xml`);
    return latest;
  }
  return null;
}

/**
 * Convert a single rss-parser item into a normalized Opportunity record.
 * @param {object} item   rss-parser feed item
 * @param {object} cfg    source config
 * @param {string} now    ISO timestamp string
 * @returns {object}
 */
function parseItem(item, cfg, now) {
  const source = cfg.source;
  const url    = item.link || '';
  const title  = item.title || '';

  // Summary — prefer full content, fall back to snippet
  const summaryRaw = item['content:encoded'] || item.content || item.contentSnippet || null;
  const summary    = cleanHtml(summaryRaw);

  // Date — rss-parser exposes isoDate when the feed has a valid date
  const publishedAt = item.isoDate
    ? new Date(item.isoDate).toISOString().replace(/\.\d{3}Z$/, 'Z')
    : null;

  // Tags: merge source defaults with feed categories
  const tags = [...(cfg.default_tags || [])];
  if (Array.isArray(item.categories)) {
    for (const cat of item.categories) {
      const term = typeof cat === 'string' ? cat : (cat && cat._ );
      if (term && !tags.includes(term)) tags.push(term);
    }
  }

  return {
    id:         sha1Id(source, url),
    title,
    url,
    source,
    source_url:   cfg.source_url,
    published_at: publishedAt,
    summary,
    tags,
    location:   null,
    deadline:   null,
    language:   cfg.language || null,
    created_at: now,
    updated_at: now,
  };
}

/**
 * Parse the raw feed for one source and return an array of Opportunity records.
 * @param {object} cfg
 * @param {string} today
 * @param {string} now
 * @returns {Promise<object[]>}
 */
async function normalizeSource(cfg, today, now) {
  const { source } = cfg;
  const rawPath = findRawFile(source, today);
  if (!rawPath) {
    console.log(`  ✗ No raw file found for source '${source}' — skipping.`);
    return [];
  }

  console.log(`  → Parsing ${rawPath} …`);
  const xmlContent = fs.readFileSync(rawPath, 'utf8');

  const parser = new RssParser({
    customFields: {
      item: [['content:encoded', 'content:encoded']],
    },
  });

  let feed;
  try {
    feed = await parser.parseString(xmlContent);
  } catch (err) {
    console.error(`  ✗ Feed parse error: ${err.message}`);
    return [];
  }

  const records = [];
  for (const item of feed.items || []) {
    try {
      const record = parseItem(item, cfg, now);
      if (record.url) records.push(record); // skip entries without a URL
    } catch (err) {
      console.error(`  ⚠ Skipping entry '${item.title || ''}' due to error: ${err.message}`);
    }
  }

  console.log(`  ✓ Parsed ${records.length} record(s) from ${source}`);
  return records;
}

// ---------------------------------------------------------------------------

async function main() {
  const sources = loadSources();
  const enabled = sources.filter(s => s.enabled);
  const today   = todayStr();
  const now     = nowIso();

  console.log(`[normalize] Date: ${today}  |  ${enabled.length} enabled source(s)`);

  const allRecords = [];

  for (const cfg of enabled) {
    console.log(`\n[${cfg.source}]`);
    try {
      const records = await normalizeSource(cfg, today, now);
      allRecords.push(...records);
    } catch (err) {
      console.error(`  ✗ Unexpected error for '${cfg.source}': ${err.message}`);
    }
  }

  if (allRecords.length > 0) {
    safeMkdir(NORM_DIR);
    const outPath = path.join(NORM_DIR, `${today}.jsonl`);
    writeJsonl(allRecords, outPath);
    console.log(`\n[normalize] Wrote ${allRecords.length} record(s) → ${outPath}`);
  } else {
    console.log('\n[normalize] No records to write.');
  }
}

main().catch(err => {
  console.error('[normalize] Fatal:', err.message);
  process.exit(1);
});

File: scripts/normalize.py
"""
normalize.py — Parse today's raw XML feed for each enabled source and write
normalized Opportunity records to data/normalized/opportunities/<today>.jsonl.

Usage:
    python scripts/normalize.py
"""

import html
import json
import os
import sys

import feedparser  # type: ignore

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.dirname(SCRIPT_DIR)

sys.path.insert(0, SCRIPT_DIR)
from utils import (  # noqa: E402
    now_iso,
    parse_date_to_iso,
    safe_mkdir,
    sha1_id,
    today_str,
    write_jsonl,
)

SOURCES_FILE = os.path.join(REPO_ROOT, "data", "sources", "sources.json")
RAW_DIR = os.path.join(REPO_ROOT, "data", "raw")
NORM_DIR = os.path.join(REPO_ROOT, "data", "normalized", "opportunities")


def load_sources() -> list[dict]:
    with open(SOURCES_FILE, "r", encoding="utf-8") as fh:
        return json.load(fh)


def find_raw_file(source: str, today: str) -> str | None:
    """Return the path to today's raw XML, falling back to latest.xml."""
    source_dir = os.path.join(RAW_DIR, source)
    daily = os.path.join(source_dir, f"{today}.xml")
    if os.path.exists(daily):
        return daily
    latest = os.path.join(source_dir, "latest.xml")
    if os.path.exists(latest):
        print(f"  ⚠ No daily file for {today}, falling back to latest.xml")
        return latest
    return None


def clean_html(text: str | None) -> str | None:
    """Strip HTML tags from *text* and unescape HTML entities."""
    if not text:
        return None
    import re

    text = re.sub(r"<[^>]+>", " ", text)
    text = html.unescape(text)
    return " ".join(text.split()) or None


def parse_entry(entry: object, cfg: dict, now: str) -> dict:
    """Convert a feedparser entry into a normalized Opportunity dict."""
    source = cfg["source"]
    url = getattr(entry, "link", "") or ""
    title = getattr(entry, "title", "") or ""

    # Summary: prefer summary, fallback to description content
    summary_raw = getattr(entry, "summary", None) or getattr(
        entry, "description", None
    )
    # Some feeds store content in entry.content list
    if not summary_raw:
        content_list = getattr(entry, "content", [])
        if content_list:
            summary_raw = content_list[0].get("value", "")
    summary = clean_html(summary_raw)

    published_at = parse_date_to_iso(getattr(entry, "published_parsed", None))
    if published_at is None:
        published_at = parse_date_to_iso(getattr(entry, "updated_parsed", None))

    tags = list(cfg.get("default_tags", []))
    # Merge feed-level tags/categories
    for tag_obj in getattr(entry, "tags", []):
        term = getattr(tag_obj, "term", None)
        if term and term not in tags:
            tags.append(term)

    return {
        "id": sha1_id(source, url),
        "title": title,
        "url": url,
        "source": source,
        "source_url": cfg["source_url"],
        "published_at": published_at,
        "summary": summary,
        "tags": tags,
        "location": None,
        "deadline": None,
        "language": cfg.get("language"),
        "created_at": now,
        "updated_at": now,
    }


def normalize_source(cfg: dict, today: str, now: str) -> list[dict]:
    """Parse the raw feed for *cfg* and return a list of Opportunity dicts."""
    source = cfg["source"]
    raw_path = find_raw_file(source, today)
    if raw_path is None:
        print(f"  ✗ No raw file found for source '{source}' — skipping.")
        return []

    print(f"  → Parsing {raw_path} …")
    feed = feedparser.parse(raw_path)

    if feed.bozo and not feed.entries:
        print(f"  ✗ Feed parse error: {feed.bozo_exception}")
        return []

    records = []
    for entry in feed.entries:
        try:
            record = parse_entry(entry, cfg, now)
            if record["url"]:  # skip entries without a URL
                records.append(record)
        except Exception as exc:  # noqa: BLE001
            entry_url = getattr(entry, "link", "<no url>")
            entry_title = getattr(entry, "title", "<no title>")
            print(f"  ⚠ Skipping entry '{entry_title}' ({entry_url}) due to error: {exc}")

    print(f"  ✓ Parsed {len(records)} record(s) from {source}")
    return records


def main() -> None:
    sources = load_sources()
    enabled = [s for s in sources if s.get("enabled", False)]
    today = today_str()
    now = now_iso()

    print(f"[normalize] Date: {today}  |  {len(enabled)} enabled source(s)")

    all_records: list[dict] = []
    for cfg in enabled:
        print(f"\n[{cfg['source']}]")
        records = normalize_source(cfg, today, now)
        all_records.extend(records)

    if all_records:
        safe_mkdir(NORM_DIR)
        out_path = os.path.join(NORM_DIR, f"{today}.jsonl")
        write_jsonl(all_records, out_path)
        print(f"\n[normalize] Wrote {len(all_records)} record(s) → {out_path}")
    else:
        print("\n[normalize] No records to write.")


if __name__ == "__main__":
    main()
